### PR TITLE
[LEC-11] Splitted interfaces and Validation Handler 

### PR DESCRIPTION
**Pull request by Horchynsky**

### CODE REVIEW

File: src/main/java/com/eleks/academy/whoami/configuration/GameControllerAdvice.java [hunk 1]
```diff
@@ -1,19 +1,40 @@
...
 	@ResponseStatus(HttpStatus.BAD_REQUEST)
 	public ApiError handleGameException(GameException gameException) {
 		return gameException::getMessage;
 	}
 
+	@Override
+	protected ResponseEntity<Object> handleMethodArgumentNotValid(MethodArgumentNotValidException ex,
+																  HttpHeaders headers, HttpStatus status,
+																  WebRequest request) {
+		List<String> details = new ArrayList<>();
```
Comment (sergvasylchak) [inline src/main/java/com/eleks/academy/whoami/configuration/GameControllerAdvice.java:33]: This method can be performed in 1 stream.
There is a `ResponseEntity.badRequest()` method to omit `HttpStatus.BAD_REQUEST` uasge.

```
	return ex.getBindingResult().getAllErrors()
				.stream()
				.map(ObjectError::getDefaultMessage)
				.collect(collectingAndThen(
						toList(),
						details -> ResponseEntity.badRequest().body(new ErrorResponse("Validation failed!", details))
				));
```

File: src/main/java/com/eleks/academy/whoami/controller/GameController.java [hunk 2]
```diff
@@ -43,14 +51,14 @@ public ResponseEntity<GameDetails> findById(@PathVariable("id") String id,
 
 	// TODO: Should return enrolled player
 	@PostMapping("/{id}/players")
-	@ResponseStatus(HttpStatus.NO_CONTENT)
+	@ResponseStatus(HttpStatus.OK)
```
Comment (sergvasylchak) [inline src/main/java/com/eleks/academy/whoami/controller/GameController.java:54]: This is a default response status. Please drop `@ResponseStatus` at all.

File: src/main/java/com/eleks/academy/whoami/core/impl/PersistentGame.java [hunk 3]
```diff
@@ -34,6 +36,13 @@ public PersistentGame(String hostPlayer, Integer maxPlayers) {
 		this.turns.add(GameState.start(hostPlayer, maxPlayers));
 	}
 
+	@Override
+	public Optional<SynchronousPlayer> findPlayer(String player) {
+		Optional<GameState> peek = Optional.ofNullable(this.turns.peek());
```
Comment (sergvasylchak) [inline src/main/java/com/eleks/academy/whoami/core/impl/PersistentGame.java:41]: `this.applyIfPresent(this.turns.peek(), gameState -> gameState.findPlayer(player));`

File: src/test/java/com/eleks/academy/whoami/core/impl/PersistentPlayerTest.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+	@Test
+	void allowToSuggestCharacterOnlyOnce() {
+		PersistentPlayer player = new PersistentPlayer("PLayerName");
+		Future<String> character = player.suggestCharacter();
+		assertFalse(character.isDone());
+		player.setCharacter("character");
+		assertTrue(character.isDone());
+		assertThrows(IllegalStateException.class, () -> player.setCharacter("character"));
+	}
+}
```
Comment (sergvasylchak) [inline src/test/java/com/eleks/academy/whoami/core/impl/PersistentPlayerTest.java:22]: Please add a new line here =)

File: src/main/java/com/eleks/academy/whoami/core/state/AbstractGameState.java [hunk 5]
```diff
@@ -8,8 +8,8 @@
 public abstract sealed class AbstractGameState implements GameState
 		permits SuggestingCharacters, WaitingForPlayers, ProcessingQuestion {
 
-	private final Integer playersInGame;
```
Comment (sergvasylchak) [inline src/main/java/com/eleks/academy/whoami/core/state/AbstractGameState.java:11]: Why? Having a `@Getter` defined you will never need these references

File: src/main/java/com/eleks/academy/whoami/core/state/WaitingForPlayers.java [hunk 6]
```diff
@@ -1,28 +1,26 @@
...
-		super(null, maxPlayers);
+	public WaitingForPlayers(int maxPlayers) {
+		super(0, maxPlayers);
+		this.players = new HashMap<>(maxPlayers);
+	}
 
-		this.maxPlayers = maxPlayers;
-		this.players = new HashMap<>(this.maxPlayers);
+	private WaitingForPlayers(int maxPlayers, Map<String, SynchronousPlayer> players) {
+		super(0, maxPlayers);
```
Comment (sergvasylchak) [inline src/main/java/com/eleks/academy/whoami/core/state/WaitingForPlayers.java:22]: Are you sure it's gonna be 0 and not `players.size()` here?

File: src/main/java/com/eleks/academy/whoami/configuration/GameControllerAdvice.java (hunk 1 above)
Comment (Horchynsky) [reply src/main/java/com/eleks/academy/whoami/configuration/GameControllerAdvice.java]: done

File: src/main/java/com/eleks/academy/whoami/controller/GameController.java (hunk 2 above)
Comment (Horchynsky) [reply src/main/java/com/eleks/academy/whoami/controller/GameController.java]: done

File: src/main/java/com/eleks/academy/whoami/core/impl/PersistentGame.java (hunk 3 above)
Comment (Horchynsky) [reply src/main/java/com/eleks/academy/whoami/core/impl/PersistentGame.java]: done

File: src/test/java/com/eleks/academy/whoami/core/impl/PersistentPlayerTest.java (hunk 4 above)
Comment (Horchynsky) [reply src/test/java/com/eleks/academy/whoami/core/impl/PersistentPlayerTest.java]: done

File: src/main/java/com/eleks/academy/whoami/core/state/AbstractGameState.java (hunk 5 above)
Comment (Horchynsky) [reply src/main/java/com/eleks/academy/whoami/core/state/AbstractGameState.java:11]: done

File: src/main/java/com/eleks/academy/whoami/core/state/WaitingForPlayers.java (hunk 6 above)
Comment (Horchynsky) [reply src/main/java/com/eleks/academy/whoami/core/state/WaitingForPlayers.java]: done

**Review by sergvasylchak — APPROVED** [review]

**Pull request closed without merging**